pkg/trace/writer/newwriter/trace.go

```diff
@@ -0,0 +1,263 @@
+const pathTraces = "/api/v0.2/traces"
+
+// payloadFlushThreshold specifies the maximum accumulated payload size that is allowed before
+// a flush is triggered; replaced in tests.
+var payloadFlushThreshold = 3200000 // 3.2MB is the maximum allowed by the Datadog API
+
+// SampledSpans represents the result of a trace sampling operation.
+type SampledSpans struct {
+	// Trace will contain a trace if it was sampled or be empty if it wasn't.
+	Trace pb.Trace
+	// Events contains all APM events extracted from a trace. If no events were extracted, it will be empty.
+	Events []*pb.Span
+}
+
+// Empty returns true if this TracePackage has no data.
+func (ss *SampledSpans) Empty() bool {
+	return len(ss.Trace) == 0 && len(ss.Events) == 0
+}
+
+// size returns the estimated size of the package.
+func (ss *SampledSpans) size() int {
+	// we use msgpack's Msgsize() heuristic because it is a good indication
+	// of the weight of a span and the msgpack size is relatively close to
+	// the protobuf size, which is expensive to compute.
+	return ss.Trace.Msgsize() + pb.Trace(ss.Events).Msgsize()
+}
+
+type TraceWriter struct {
+	in       <-chan *SampledSpans
+	hostname string
+	env      string
+	senders  []*sender
+	stop     chan struct{}
+
+	stats *info.TraceWriterInfo
+
+	traces       []*pb.APITrace // traces buffered
+	events       []*pb.Span     // events buffered
+	bufferedSize int            // estimated buffer size
+}
+
+func NewTraceWriter(cfg *config.AgentConfig, in <-chan *SampledSpans) *TraceWriter {
+	// allow 10% of the connection limit to outgoing sends.
+	climit := int(math.Max(1, float64(cfg.ConnectionLimit)/10))
+	tw := &TraceWriter{
+		in:       in,
+		hostname: cfg.Hostname,
+		env:      cfg.DefaultEnv,
+		stats:    &info.TraceWriterInfo{},
+		stop:     make(chan struct{}),
+	}
+	tw.senders = newSenders(cfg, tw, pathTraces, climit)
+	return tw
+}
+
+// Stop stops the TraceWriter and attempts to flush whatever is left in the senders buffers.
+func (w *TraceWriter) Stop() {
+	log.Debug("Exiting trace writer. Trying to flush whatever is left...")
+	w.stop <- struct{}{}
+	<-w.stop
+	var wg sync.WaitGroup
+	for _, s := range w.senders {
+		wg.Add(1)
+		go func(s *sender) {
+			defer wg.Done()
+			s.Flush()
+		}(s)
+	}
+	wg.Wait()
+}
+
+// Run starts the TraceWriter.
+func (w *TraceWriter) Run() {
+	log.Debug("Starting trace writer")
+	t := time.NewTicker(5 * time.Second)
+	defer t.Stop()
+	defer close(w.stop)
+	for {
+		select {
+		case pkg := <-w.in:
+			w.addSpans(pkg)
+		case <-w.stop:
+			// drain the input channel before stopping
+		outer:
+			for {
+				select {
+				case pkg := <-w.in:
+					w.addSpans(pkg)
+				default:
+					break outer
+				}
+			}
+			w.flush()
+			return
+		case <-t.C:
+			w.report()
+			w.flush()
+		}
+	}
+}
+
+func (w *TraceWriter) addSpans(pkg *SampledSpans) {
+	if pkg.Empty() {
+		return
+	}
+
+	atomic.AddInt64(&w.stats.Spans, int64(len(pkg.Trace)+len(pkg.Events)))
```

Comment:
are these counts overlapping ?

Reply:
Not exactly sure what you mean. These are metrics we've always had. Tried to keep the same.

Reply:
Maybe we should measure events separately 🤔I will add another metric. It seems misleading to include events into "spans"